spinalcordtoolbox/utils/shell.py

```diff
@@ -98,6 +99,38 @@ def display_viewer_syntax(files, colormaps=[], minmax=[], opacities=[], mode='',
         printv(cmd + '\n', verbose=1, type='info')
 
 
+class SCTArgumentParser(argparse.ArgumentParser):
+    """Parser that centralizes initialization steps common across all SCT scripts."""
+    def __init__(self, *args, **kwargs):
+        def update_parent_default(key, value):
+            """A polite way of letting a child class have different default values than the parent class."""
+            # Source: https://stackoverflow.com/a/41623488
+            argspec = inspect.getfullargspec(super(SCTArgumentParser, self).__init__)
+            arg_index = argspec.args.index(key)
+            if len(args) < arg_index and key not in kwargs:
+                kwargs[key] = value
+
+        update_parent_default('formatter_class', SmartFormatter)
+
+        # Update "usage:" message to match how SCT scripts are actually called (no '.py')
+        frame = inspect.stack()[1]
+        module = inspect.getmodule(frame[0])
+        update_parent_default('prog', os.path.basename(module.__file__).strip(".py"))
+
+        # Disable "add_help", because it won't properly add '-h' to our custom argument groups
+        # (We use custom argument groups because of https://stackoverflow.com/a/24181138)
+        update_parent_default('add_help', False)
+
+        super(SCTArgumentParser, self).__init__(*args, **kwargs)
```

Comment:
> Or maybe we should just say "our custom formatter" and assume people will go read its source code if they need to figure out what it can do?

Exactly. `SmartFormatter` and `SCTArgumentParser` are in the same `utils/shell.py` file, so all someone has to do is scroll down a little.

Perhaps `utils/shell.py` could be reordered so that `SCTArgumentParser` is right next to `SmartFormatter`? The ordering is a bit haphazard right now.

Reply:
> Instead of that stackoverflow thing I think you can use the (new-ish) dict-splat syntax: https://stackoverflow.com/questions/38987/how-do-i-merge-two-dictionaries-in-a-single-expression-in-python-taking-union-o#26853961

The StackOverflow pattern checks whether the argument was passed not just via `kwargs` but also `args` (which is the trickier one). 

With the dict-splat method, if someone passed an argument via `args`, I think our default will still get added to `kwargs` (i.e. that argument it will be passed twice with possibly conflicting values). I'm honestly not sure what the result is when that happens... hrm.

Reply:
Yeah, I get an error when I try the suggestion. :cry:  This part of the test:

```python
    # Check that new defaults can still be overridden (setting add_help via args AND kwargs)
    parser1 = utils.SCTArgumentParser(None, None, None, None, [], utils.SmartFormatter, '-', None, None, 'error', True)
    assert parser1.add_help is True
```

Results in this error:

```python
>       super(SCTArgumentParser, self).__init__(*args, **{**defaults, **kwargs})
E       TypeError: __init__() got multiple values for argument 'formatter_class'
```

Since it gets passed in both `args` and `defaults`.

Reply:
I didn't think about that at all. Thanks for trying it out. Now I've learned something.

This seems like it should be solvable with shorter syntax but :shrug: it's not a big deal. And you documented it well so it won't be a big mystery in the future.